File: src/main/scala/actor/MongoCollectionActor.scala
package actor

import akka.actor.Actor
import util.Configuration
import com.mongodb.casbah.Imports._


trait MongoCollectionActor extends Actor {

  private lazy val db = Configuration.db

  protected lazy val coll = db(this.name)

  protected val name: String

  def indexCollection()


  def receive = {
    case Write(objects: List[MongoDBObject]) => {
      objects.map(obj => coll += obj)
      sender tell(Done(s"$name"), self)
    }

    case FindOne(query: MongoDBObject) => {
      val result = coll.findOne(query)
      sender tell(result, self)
    }

    case Find(query: MongoDBObject) => {
      sender tell(coll.find(query), self)
    }
  }


  override def preStart() {
    super.preStart()
    println(s"Preparing collection $name")

    coll.drop()
    coll.dropIndexes()

    indexCollection()
  }

  override def postStop() {
    super.postStop()
    println(s"Collection $name:. ${coll.count()} elements")
  }
}

File: src/main/java/actor/MongoCollectionActor.java
package actor;

import actor.message.Done;
import actor.message.Find;
import actor.message.FindOne;
import actor.message.Write;
import akka.actor.UntypedActor;
import com.mongodb.DB;
import org.jongo.Jongo;
import org.jongo.MongoCollection;
import util.Configuration;

import java.net.UnknownHostException;

public abstract class MongoCollectionActor<T> extends UntypedActor {

    private MongoCollection collection;

    protected abstract String getCollectionName();

    protected abstract void indexCollection(MongoCollection mongoCollection);


    @Override
    public void onReceive(Object message) throws Exception {
        if (message instanceof Write) {
            collection.insert(((Write) message).objects);
            getSender().tell(new Done(getCollectionName()));
        } else {
            tryFind(message);
        }
    }

    private void tryFind(Object message) {
        if (message instanceof Find) {
            Find<T> find = (Find) message;
            getSender().tell(collection.find(find.query).as(find.clazz));
        } else if (message instanceof FindOne) {
            FindOne<T> findOne = (FindOne<T>) message;
            getSender().tell(collection.findOne(findOne.query).as(findOne.clazz));
        } else {
            unhandled(message);
        }
    }

    @Override
    public void preStart() {
        super.preStart();

        try {
            DB db = Configuration.getDb();
            Jongo jongo = new Jongo(db);
            collection = jongo.getCollection(getCollectionName());
            collection.drop();
            indexCollection(collection);

        } catch (UnknownHostException e) {
            throw new IllegalStateException(e);
        }
    }
}
